src/widget/switch.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright 2022 The Druid Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use glazier::kurbo::Circle;
+use vello::{
+    kurbo::{Point, Size},
+    peniko::Color,
+    SceneBuilder,
+};
+
+use crate::{IdPath, Message};
+
+use super::{
+    contexts::LifeCycleCx,
+    piet_scene_helpers::{fill_color, stroke},
+    AccessCx, BoxConstraints, ChangeFlags, Event, EventCx, LayoutCx, LifeCycle, PaintCx, UpdateCx,
+    Widget,
+};
+
+pub struct Switch {
+    id_path: IdPath,
+    is_on: bool,
+    is_moved: bool,
+    knob_position: Point,
+}
+
+impl Switch {
+    pub fn new(id_path: &IdPath, is_on: bool) -> Switch {
+        Switch {
+            id_path: id_path.clone(),
+            is_on,
+            is_moved: false,
+            knob_position: if is_on {
+                Point::new(ON_POS, KNOB_DIAMETER / 2. + SWITCH_PADDING)
+            } else {
+                Point::new(OFF_POS, KNOB_DIAMETER / 2. + SWITCH_PADDING)
+            },
+        }
+    }
+
+    pub fn set_is_on(&mut self, is_on: bool) -> ChangeFlags {
+        self.is_on = is_on;
+        ChangeFlags::PAINT
+    }
+}
+
+// See druid's button for info.
+const KNOB_DIAMETER: f64 = 20.0;
+const SWITCH_PADDING: f64 = 3.0;
+const SWITCH_WIDTH: f64 = 2.0 * KNOB_DIAMETER + 2.0 * SWITCH_PADDING;
+const SWITCH_HEIGHT: f64 = KNOB_DIAMETER + 2.0 * SWITCH_PADDING;
+const ON_POS: f64 = SWITCH_WIDTH - KNOB_DIAMETER / 2.0 - SWITCH_PADDING;
+const OFF_POS: f64 = KNOB_DIAMETER / 2.0 + SWITCH_PADDING;
+
+impl Widget for Switch {
+    fn event(&mut self, cx: &mut EventCx, event: &Event) {
+        match event {
+            Event::MouseDown(_) => {
+                cx.set_active(true);
+                cx.request_paint();
+            }
+            Event::MouseUp(_) => {
+                if self.is_moved {
+                    if self.is_on != (self.knob_position.x > SWITCH_WIDTH / 2.0) {
+                        cx.add_message(Message::new(self.id_path.clone(), ()))
+                    }
+                } else if cx.is_active() {
```

Comment:
(Also super nit-picky, because it shouldn't really change the behavior, as a `MouseMove` event should happen before)
```suggestion
                } else if cx.is_active() && cx.is_hot() {
```

Reply:
I am replying here even though I will be creating a new PR. The is_hot check I think is wrong here and the only reason it doesn't affect the behaviour is because no actual message is being send here. Had there been any message being sent to the view it would not have if the user moved past the boundaries of the widget. Good UX would be to ensure things function correctly even in light of errors from the user

Reply:
Hmm I guess it really depends. A button is not clicked when the user releases it outside the button area. On touchscreens I agree with you though, or generally when the switch is dragged, but this should never be the case here, since `is_dragging` will be set before, so it doesn't really matter...